### ISSUE TITLE
Measure and document baseline matching accuracy across all layers

### ISSUE THREAD

**Issue by touchthesun** [helpfulengineering/supply-graph-ai#70]
Establish comprehensive baseline metrics for matching accuracy across all matching layers (Direct, Heuristic, NLP, LLM). These metrics will serve as the foundation for measuring improvements and identifying areas that need optimization.

The matching service uses a progressive layered approach:
1. **Direct Matching**: Exact string matching with near-miss detection
2. **Heuristic Matching**: Rule-based matching using capability-centric rules
3. **NLP Matching**: Semantic similarity using spaCy
4. **LLM Matching**: AI-powered advanced reasoning

We need to measure accuracy for each layer independently and for the combined system to understand where matching works well and where it needs improvement.

## Acceptance Criteria

- [ ] Accuracy metrics defined and documented:
  - [ ] True Positive Rate (Recall/Sensitivity)
  - [ ] True Negative Rate (Specificity)
  - [ ] Precision (Positive Predictive Value)
  - [ ] F1 Score (harmonic mean of precision and recall)
  - [ ] Accuracy (overall correctness)
  - [ ] Confidence score calibration
- [ ] Ground truth dataset created:
  - [ ] At least 100 requirement-capability pairs
  - [ ] Manual validation of correct matches
  - [ ] Labeled with expected match result (yes/no)
  - [ ] Coverage across all capability types
  - [ ] Edge cases included
- [ ] Baseline measurements for each layer:
  - [ ] Direct matching accuracy documented
  - [ ] Heuristic matching accuracy documented
  - [ ] NLP matching accuracy documented
  - [ ] LLM matching accuracy documented (if available)
  - [ ] Combined system accuracy documented
- [ ] Per-capability-type breakdown:
  - [ ] Accuracy by capability category (tools, materials, processes, etc.)
  - [ ] Identification of weak categories
- [ ] Confidence score analysis:
  - [ ] Correlation between confidence and accuracy
  - [ ] Calibration curves generated
  - [ ] Threshold recommendations
- [ ] Documentation:
  - [ ] Methodology documented
  - [ ] Results published in docs/metrics/
  - [ ] Comparison with target metrics (>90% for direct/heuristic)
- [ ] Automated accuracy testing:
  - [ ] Test suite for continuous accuracy monitoring
  - [ ] Integration with CI/CD
  - [ ] Regression alerts

## Dependencies

- Issue 1.1.2 (test dataset provides test cases)

## Estimated Effort

L (6-8 days)

## Technical Notes

Create ground truth dataset in `tests/data/matching/ground_truth/`:
```json
{
  "test_cases": [
    {
      "id": "match-001",
      "requirement": {
        "type": "tool",
        "value": "3D printer",
        "context": "professional"
      },
      "capability": {
        "type": "tool",
        "value": "FDM 3D Printer (Prusa i3 MK3S+)",
        "specifications": {...}
      },
      "expected_match": true,
      "confidence_threshold": 0.8,
      "notes": "Direct tool type match"
    }
  ]
}
```

Implement accuracy measurement framework:
```python
from dataclasses import dataclass
from typing import List, Dict
import numpy as np
from sklearn.metrics import precision_recall_fscore_support, roc_auc_score

@dataclass
class MatchResult:
    requirement_id: str
    capability_id: str
    matched: bool
    confidence: float
    layer: str  # 'direct', 'heuristic', 'nlp', 'llm'

@dataclass
class AccuracyMetrics:
    precision: float
    recall: float
    f1_score: float
    accuracy: float
    true_positives: int
    false_positives: int
    true_negatives: int
    false_negatives: int
    confidence_auc: float  # AUC for confidence scores

class MatchingAccuracyEvaluator:
    def __init__(self, ground_truth_path: str):
        self.ground_truth = self.load_ground_truth(ground_truth_path)
    
    def evaluate_layer(self, layer: str, results: List[MatchResult]) -> AccuracyMetrics:
        """Evaluate accuracy for a specific matching layer"""
        y_true = [gt.expected_match for gt in self.ground_truth]
        y_pred = [r.matched for r in results]
        y_confidence = [r.confidence for r in results]
        
        precision, recall, f1, _ = precision_recall_fscore_support(
            y_true, y_pred, average='binary'
        )
        
        # Calculate confusion matrix elements
        tp = sum(1 for t, p in zip(y_true, y_pred) if t and p)
        fp = sum(1 for t, p in zip(y_true, y_pred) if not t and p)
        tn = sum(1 for t, p in zip(y_true, y_pred) if not t and not p)
        fn = sum(1 for t, p in zip(y_true, y_pred) if t and not p)
        
        accuracy = (tp + tn) / (tp + fp + tn + fn)
        confidence_auc = roc_auc_score(y_true, y_confidence)
        
        return AccuracyMetrics(
            precision=precision,
            recall=recall,
            f1_score=f1,
            accuracy=accuracy,
            true_positives=tp,
            false_positives=fp,
            true_negatives=tn,
            false_negatives=fn,
            confidence_auc=confidence_auc
        )
    
    def generate_report(self, results_by_layer: Dict[str, List[MatchResult]]):
        """Generate comprehensive accuracy report"""
        report = {
            "timestamp": datetime.now().isoformat(),
            "layers": {}
        }
        
        for layer, results in results_by_layer.items():
            metrics = self.evaluate_layer(layer, results)
            report["layers"][layer] = {
                "precision": metrics.precision,
                "recall": metrics.recall,
                "f1_score": metrics.f1_score,
                "accuracy": metrics.accuracy,
                "confusion_matrix": {
                    "tp": metrics.true_positives,
                    "fp": metrics.false_positives,
                    "tn": metrics.true_negatives,
                    "fn": metrics.false_negatives
                },
                "confidence_auc": metrics.confidence_auc
            }
        
        return report
```

Generate calibration curves for confidence scores:
```python
import matplotlib.pyplot as plt
from sklearn.calibration import calibration_curve

def plot_calibration_curve(y_true, y_confidence, layer_name):
    """Plot reliability diagram for confidence calibration"""
    fraction_of_positives, mean_predicted_value = calibration_curve(
        y_true, y_confidence, n_bins=10
    )
    
    plt.figure(figsize=(10, 6))
    plt.plot(mean_predicted_value, fraction_of_positives, "s-", label=layer_name)
    plt.plot([0, 1], [0, 1], "k--", label="Perfect calibration")
    plt.xlabel("Mean predicted confidence")
    plt.ylabel("Fraction of positives")
    plt.title(f"Calibration Curve - {layer_name}")
    plt.legend()
    plt.grid()
    plt.savefig(f"docs/metrics/calibration_{layer_name}.png")
```

Store baseline results in `docs/metrics/baseline_accuracy.json`:
```json
{
  "version": "0.9.0",
  "timestamp": "2026-01-22T00:00:00Z",
  "target_metrics": {
    "direct_matching": {"precision": 0.95, "recall": 0.90, "f1": 0.92},
    "heuristic_matching": {"precision": 0.90, "recall": 0.85, "f1": 0.87},
    "nlp_matching": {"precision": 0.85, "recall": 0.80, "f1": 0.82},
    "combined": {"precision": 0.90, "recall": 0.90, "f1": 0.90}
  },
  "baseline_metrics": {
    "direct_matching": {...},
    "heuristic_matching": {...},
    "nlp_matching": {...},
    "combined": {...}
  }
}
```

---
*Generated from `notes/github-issues.md` — Issue 1.2.1*
